### DISCUSSION TITLE
dataclass support for parameters in dependencies

### DISCUSSION THREAD

**Discussion by dz0gchen** [fastapi/fastapi#8795] [Questions]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [ ] I commit to help with one of those options 👆

### Example Code

```python
@dataclasses
class MaterialParams:
        ...
        

async def get_articles(
    params: MaterialParams = Depends(MaterialParams),
):
```

### Description

need dataclass support for parameters in dependencies 

### Wanted Solution

dataclass support for parameters in dependencies 

### Wanted Code

```python
@dataclasses
class MaterialParams:
        ...
        

async def get_articles(
    params: MaterialParams = Depends(MaterialParams),
):
```

### Operating System

Linux

### FastAPI Version

0.68.0

### Python Version

3.9

**Comment by lucastosetto** [comment]
I'm working on this one!

**Comment by lucastosetto** [comment]
@dz0gchen 

After running some tests I found out that the issue with dataclasses happens when the class is unhashable (that is, does not have a `__hash__()` method). By default `@dataclass` decorator will treat this as described [here](https://docs.python.org/3/library/dataclasses.html) (search for `unsafe_hash` paragraph).

With that being said, a workaround is to replace `@dataclass` by `@dataclass(eq=True, frozen=True)` (I tested an it works just like a "non-dataclass" dependency). This means an __eq__() method will be generated (`eq=True`) and assigning to fields in your class will generate an exception (`frozen=True`), but it ensures that the `__hash__()` method will still only be created if it is safe to do so.

Another approach, which I haven't personally tested, is to use `@dataclass(unsafe_hash=True)`. But, as the name says itself, it may be unsafe...

Because of that behaviour, the only way we can ensure a dependency will always support dataclass is to have the second approach enforced inside FastAPI, which is clearly not a good practice.

I think it's nice to have this as a note on the docs though.
